src/include/duckdb/execution/index/fixed_size_allocator.hpp

```diff
@@ -0,0 +1,125 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/execution/index/fixed_size_allocator.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/types/validity_mask.hpp"
+#include "duckdb/common/unordered_set.hpp"
+#include "duckdb/storage/buffer_manager.hpp"
+#include "duckdb/storage/metadata/metadata_manager.hpp"
+#include "duckdb/storage/metadata/metadata_writer.hpp"
+#include "duckdb/execution/index/fixed_size_buffer.hpp"
+#include "duckdb/execution/index/index_pointer.hpp"
+#include "duckdb/common/unordered_map.hpp"
+#include "duckdb/common/constants.hpp"
+
+namespace duckdb {
+
+//! The FixedSizeAllocator provides pointers to fixed-size memory segments of pre-allocated memory buffers.
+//! The pointers are IndexPointers, and the leftmost byte (metadata) must always be zero.
+//! It is also possible to directly request a C++ pointer to the underlying segment of an index pointer.
+class FixedSizeAllocator {
+public:
+	//! We can vacuum 10% or more of the total in-memory footprint
+	static constexpr uint8_t VACUUM_THRESHOLD = 10;
+
+	//! Constants for fast offset calculations in the bitmask
+	static constexpr idx_t BASE[] = {0x00000000FFFFFFFF, 0x0000FFFF, 0x00FF, 0x0F, 0x3, 0x1};
+	static constexpr uint8_t SHIFT[] = {32, 16, 8, 4, 2, 1};
```

Comment:
For a later PR:

We can make the ART fully buffered managed now by unpinning the buffers of the fixed size allocator after we are done with an operation, e.g. we can `Unpin` during/after the vacuum, which then allows us to evict the ART blocks if required because of other memory pressure. This will allow e.g. group bys after inserts to a table with a primary key to reclaim the memory used by the ART and avoid permanently keeping the ART taking up space in memory.